lib/unifex/code_generator/base_types/struct_template.ex

```diff
@@ -0,0 +1,269 @@
+        ERL_NIF_TERM value;
+
+        #{fields_parsing}
+        #{result};
+      })
+      """
+    end
+  end
+
+  defmodule CNode do
+    use Unifex.CodeGenerator.BaseType
+    alias Unifex.CodeGenerator.BaseType
+
+    def generate_arg_serialize(name, ctx, struct_module_name, struct_fields) do
+      fields_serialization =
+        struct_fields
+        |> Enum.map(fn {field_name, field_type} ->
+          ~g"""
+          ei_x_encode_atom(out_buff, "#{field_name}");
+          #{BaseType.generate_arg_serialize(field_type, :"#{name}.#{field_name}", ctx.generator)};
+          """
+        end)
+        |> Enum.join("\n")
+
+      ~g"""
+      ({
+        ei_x_encode_map_header(out_buff, #{length(struct_fields) + 1});
+        #{fields_serialization}
+        ei_x_encode_atom(out_buff, "__struct__");
+        ei_x_encode_atom(out_buff, "Elixir.#{struct_module_name |> Atom.to_string()}");
+      });
+      """
+    end
+
+    def generate_arg_parse(arg, var_name, ctx, struct_fields) do
+      %{postproc_fun: postproc_fun, generator: generator} = ctx
+
+      fields_parsing =
+        struct_fields
+        |> Enum.map(fn {field_name, field_type} ->
+          ~g"""
+          if (strcmp(key, "#{field_name}") == 0) {
+            #{
+            BaseType.generate_arg_parse(
+              field_type,
+              :"#{var_name}.#{field_name}",
+              arg,
+              postproc_fun,
+              generator
+            )
+          }
+          }
+          """
+        end)
+        |> Enum.concat([
+          ~g"""
+          if (strcmp(key, "__struct__") == 0) {
+            char* elixir_module_name;
+            #{
+            BaseType.generate_arg_parse(:atom, :elixir_module_name, arg, postproc_fun, generator)
+          }
+            #{BaseType.generate_destruction(:atom, :elixir_module_name, generator)}
+          }
+          """
+        ])
+        |> Enum.join(" else ")
+
+      ~g"""
+      ({
+        int arity = 0;
+        int decode_map_header_result = ei_decode_map_header(#{arg}->buff, #{arg}->index, &arity);
+        if (decode_map_header_result == 0) {
+          for (int i = 0; i < arity; ++i) {
+            char key[MAXATOMLEN + 1];
+            int decode_key_result = ei_decode_atom(#{arg}->buff, #{arg}->index, key);
+            if (decode_key_result == 0) {
+              #{fields_parsing}
+            }
+          }
+        }
+
+        decode_map_header_result;
+      })
+      """
+    end
+  end
+
+  def compile_struct_module(struct_type_name, struct_module_name, struct_fields) do
+    ignore_module_conflicts_option = Code.compiler_options()[:ignore_module_conflict] || false
+    Code.put_compiler_option(:ignore_module_conflict, true)
+
+    compiled_code =
+      quoted_module_code(struct_type_name, struct_module_name, struct_fields)
+      |> Code.compile_quoted()
+
+    Code.put_compiler_option(:ignore_module_conflict, ignore_module_conflicts_option)
+
+    compiled_code
+  end
```

Comment:
I am wondering if this is a good idea.

At this moment we create and compile separate module for each user defined struct. These modules are exactly the same, they differ only in name. I understand this approach becasue callbacks from `BaseType` are called based on variable type so e.g. for `int` we call calbacks from `BaseTypes.Int` for string we call callbacks from `BaseTypes.String` and so on. The problem is with struct becasue we resolve them by `struct_type_name` (if I understand it properly).

The better option would be to write one module called `struct` (instead of `struct_template`) and for each user defined struct call functions from this module.